clusterloader2: improve experience on aks

alexeldeib opened the pull request:
AKS, much like GKE, does not support SSH to master nodes or similar methods of contacting them. This PR avoids trying to do so as part of cluster loader tests for AKS. Without this, the TestMetrics gather step fails to SSH without a signer for the provider.

I imagine the same patterns would be broadly applicable for managed clusters. Perhaps we could generalize this?

marwanad commented on clusterloader2/pkg/chaos/nodes.go:
```diff
@@ -82,7 +82,7 @@ func (r *eventRecorder) record(a nodeAction, nodeName string) {
 // NewNodeKiller creates new NodeKiller.
 func NewNodeKiller(config api.NodeFailureConfig, client clientset.Interface, provider string) (*NodeKiller, error) {
-	if provider != "gce" && provider != "gke" {
+	if provider != "gce" && provider != "aks" {
```
why remove gke?

alexeldeib replied:
whoops

mm4tt commented on clusterloader2/pkg/measurement/common/etcd_metrics.go:
```diff
@@ -184,7 +184,8 @@ func (e *etcdMetricsMeasurement) stopAndSummarize(host, provider string, port in
-	if provider == "gke" {
+	// TODO(ace): --managed flag instead of provider switch?
+	if provider == "gke" || provider == "aks" {
 		klog.Infof("%s: not grabbing etcd metrics through master SSH: unsupported for gke", e)
```
Please change to ...unsupported for %s", e, provider)

mm4tt commented on clusterloader2/pkg/measurement/common/profile.go:
```diff
@@ -187,6 +187,11 @@ func (p *profileMeasurement) gatherProfile(c clientset.Interface) ([]measurement
 
+		// ssh to collect profile metrics not supported on aks masters
+		if p.config.provider == "aks" {
+			return nil, nil
```
It's possible to do it without SSH, see line 202. Would it be feasible for aks?

alexeldeib replied:
testing this out, I think it should work...will report back

mm4tt commented on clusterloader2/pkg/measurement/common/etcd_metrics.go:
```diff
@@ -184,7 +184,8 @@ func (e *etcdMetricsMeasurement) stopAndSummarize(host, provider string, port in
 func (e *etcdMetricsMeasurement) getEtcdMetrics(host, provider string, port int) ([]*model.Sample, error) {
 	// Etcd is only exposed on localhost level. We are using ssh method
-	if provider == "gke" {
+	// TODO(ace): --managed flag instead of provider switch?
```
I'd rather create a new field in ClusterLoaderConfig: `isSshSupported` that would be initialized depending on the provider. Then we could have a generic condition for disabling things that require ssh

mm4tt replied:
Actually ClusterConfig would be a better place

alexeldeib replied:
moved to cluster config and used the value throughout. I left the prometheus one as individually checking the values since it's not technically the same condition -- let me know if you think some more cleanup is appropriate there.

mm4tt commented on clusterloader2/pkg/measurement/common/scheduler_latency.go:
```diff
@@ -226,7 +226,7 @@ func (s *schedulerLatencyMeasurement) sendRequestToScheduler(c clientset.Interfa
 	} else {
 		// If master is not registered fall back to old method of using SSH.
-		if provider == "gke" {
+		if provider == "gke" || provider == "aks" {
```
Same here

mm4tt commented on clusterloader2/cmd/clusterloader.go:
```diff
@@ -143,6 +143,9 @@ func completeConfig(m *framework.MultiClientSet) error {
 		}
 	}
+	if clusterLoaderConfig.ClusterConfig.Provider == "aks" || clusterLoaderConfig.ClusterConfig.Provider == "gke" {
+		clusterLoaderConfig.ClusterConfig.IsSSHSupported = false
```
Do we set it to true anywhere? :)

mm4tt commented on clusterloader2/pkg/measurement/common/etcd_metrics.go:
```diff
@@ -75,6 +75,7 @@ func (e *etcdMetricsMeasurement) Execute(config *measurement.MeasurementConfig)
 	}
 
 	etcdInsecurePort := config.ClusterFramework.GetClusterConfig().EtcdInsecurePort
+	isSSHSupported := config.ClusterFramework.GetClusterConfig().IsSSHSupported
```
Hmm, I think that if the ssh is not supported then this measurement has nothing to do. In such cases we should just log and info and return from the Execute method. Does it make sense?

mm4tt commented on clusterloader2/pkg/measurement/common/profile.go:
```diff
@@ -189,8 +189,8 @@ func (p *profileMeasurement) gatherProfile(c clientset.Interface) ([]measurement
-		if p.config.provider == "gke" || shouldGetAPIServerByK8sClient(p.config.componentName) {
-			// SSH to master is not possible in gke, but if the component is
+		if !isSSHSupported || shouldGetAPIServerByK8sClient(p.config.componentName) {
+			// SSH to master is not possible in gke/ or ks, but if the component is
```
s/ks/eks/?

But, given that we have this isSSHSupported now, maybe let's remove the specific providers from the comments. Such comments will get quickly outdated once we disabled ssh for another provider. So maybe something like:
```
If we're here it means that ssh to master is not supported in the cluster. But if the component is kube-apiserver then we can get ...
```

WDYT?

mm4tt commented on clusterloader2/pkg/measurement/common/scheduler_latency.go:
```diff
@@ -226,7 +226,7 @@ func (s *schedulerLatencyMeasurement) sendRequestToScheduler(c clientset.Interfa
 	} else {
 		// If master is not registered fall back to old method of using SSH.
-		if provider == "gke" {
+		if !isSSHSupported {
```
Similarly to the etcd measurement, this measurement will have nothing to do if `!(masterRegistered || isSshSupported)`. Can we move this check to the top of the execute method?

mm4tt commented on clusterloader2/pkg/prometheus/prometheus.go:
```diff
@@ -96,7 +96,7 @@ func NewPrometheusController(clusterLoaderConfig *config.ClusterLoaderConfig) (p
 	}
 	if _, exists := mapping["PROMETHEUS_SCRAPE_APISERVER_ONLY"]; !exists {
-		mapping["PROMETHEUS_SCRAPE_APISERVER_ONLY"] = clusterLoaderConfig.ClusterConfig.Provider == "gke"
+		mapping["PROMETHEUS_SCRAPE_APISERVER_ONLY"] = clusterLoaderConfig.ClusterConfig.Provider == "gke" || clusterLoaderConfig.ClusterConfig.Provider == "aks"
```
Yeah, this is good. It shouldn't rely on the isSSHSupported

alexeldeib commented:
I really appreciate the quick reviews 🙂 thank you!

I've gone ahead and hoisted moist of the checks earlier. It doesn't seem like the pprof approach worked on AKS, I need to figure out why though because I would expect it to. For now I've added a flag similar to ssh. Let me know what you think of this? I also corrected the initial conditions to take advantage of the bool zero value.

sorry about random json -_-

alexeldeib commented:
/test pull-perf-tests-clusterloader2

mm4tt commented on clusterloader2/pkg/measurement/common/etcd_metrics.go:
```diff
@@ -60,6 +60,13 @@ type etcdMetricsMeasurement struct {
 func (e *etcdMetricsMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	isSSHSupported := config.ClusterFramework.GetClusterConfig().IsSSHSupported
+	// Etcd is only exposed on localhost level. We are using ssh method
+	if isSSHSupported {
```
`if !isSSHSupported` ?

Also, do we need to introduce this local variable? I don't think it improves readability a lot. What about
```
if !config.ClusterFramework.GetClusterConfig().IsSSHSupported {
  ...
}
```

mm4tt commented on clusterloader2/pkg/config/cluster.go:
```diff
@@ -40,6 +40,8 @@ type ClusterConfig struct {
 	MasterName                 string
 	KubemarkRootKubeConfigPath string
 	DeleteStaleNamespaces      bool
+	IsSSHSupported             bool
```
We should make it clear that it's the SSH to master that is not supported (e.g. in GKE you can still ssh into nodes). 
`IsSSHToMasterSupported`?

mm4tt commented on clusterloader2/pkg/config/cluster.go:
```diff
@@ -40,6 +40,8 @@ type ClusterConfig struct {
 	KubemarkRootKubeConfigPath string
 	DeleteStaleNamespaces      bool
+	IsSSHSupported             bool
+	IsPprofSupported           bool
```
Similarly to the ssh we should make it clear what it means. Could you either change the name to make it more self documenting or add a comment (or both)?

alexeldeib replied:
how's that?

mm4tt commented on clusterloader2/pkg/measurement/common/scheduler_latency.go:
```diff
@@ -61,14 +61,35 @@ type schedulerLatencyMeasurement struct{}
+	isSSHSupported := config.ClusterFramework.GetClusterConfig().IsSSHSupported
+
+	c := config.ClusterFramework.GetClientSets().GetClient()
+	nodes, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
```
Hmm, this is actually a quite heavy call (especially in large clusters) we shouldn't be making it for every Execute call. I'm aware that it was already like this, but let's make it better. Would you mind moving this logic to clusterloader.go's `completeConfig` method and adding a field to the ClusterConfig? Ideally, it could be done in a separate PR. Thanks!

alexeldeib replied:
I can tackle this in a separate PR, sure. I assume it's done here because we need to check if the master is registered. I don't love that i'm adding a bunch of new fields to this struct, but I could add something like `IsMasterNodeRegistered` to cover this and set it using this logic in `completeConfig`? WDYT?

mm4tt replied:
Yeah, that sounds great, thanks!

wojtek-t replied:
Yes - let's do in a separate PR please.

mm4tt commented on clusterloader2/pkg/config/cluster.go:
```diff
@@ -40,8 +40,12 @@ type ClusterConfig struct {
+	IsSSHToMasterSupported bool
+	// IsAPIServerPprofExposed is false for some managed providers.
+	// When IsAPIServerPprofExposed is false, clusterloader will avoid using using pprof to collect metrics.
+	IsAPIServerPprofExposed bool
```
Looking at the code, I thin the question is more whether the pprof is enabled in apiserver (we're exposing in on our own in profile.go). Can we change the name to IsAPIServerPprofEnabled and change the doc to say something like
`When IsAPIServerPprofEnabled is false, clusterloader2 will avoid collecting kube-apiserver pprofs`

alexeldeib replied:
fixed 👍

alexeldeib commented:
Would you like me to squash commits before merge? I kept everything as-is until it looked good.
